Addition of new density-density operators mentioned in the discussion…

LeoGaspard opened the pull request:
Addition of NupNdn, NupNup and NdnNdn operators. Addition of the equivalent hole density-density operators handled at the compile step following the discussion https://github.com/awietek/xdiag/discussions/50.

awietek commented:
Dear Leo,

thank you very much for the pull request. Currently, according to the automated tests, the MPI code is not compiling: https://github.com/awietek/xdiag/actions/runs/14702408841/job/41256972462?pr=53

Please have a look at it and try to fix it. I will then take a closer look at the new code.

LeoGaspard commented:
Dear Alexander, 

Thank you for your input. Indeed I misunderstood what you told me, I now removed all the composed operators, leaving only NupNdn, NupNup and NdnNdn. I also added the tests in the file `test/algebra/test_apply.cpp` for the operators NupNdn (both i j and j i since it is important to check that they can be different), NupNup and NdnNdn.

awietek commented:
Thanks, now there is a test failing unfortunately with the MPI library. Could you have a look into what the problem is there?

awietek commented:
Thanks, so superficially all looks good. I will have to take a closer look now which can take one or two days, then I'll come back to you. I guess you can already use the version for project.

awietek commented on xdiag/basis/apply_identity_distributed.hpp:
This function should be moved to apply_identity.hpp. Also, this is specific to the electron bases. I would implement this more generically, something like this:

```c++
void apply_identity(Coupling const &cpl, const coeff_t *vec_in, coeff_t *vec_out, int64_t size) {
    using bit_t = typename basis_t::bit_t;
    coeff_t s = cpl.scalar().as<coeff_t>();
    for (int64_t idx=0; idx<size; ++idx) {
            vec_out[idx] += s * vec_in[idx];
    }
}
```

which will work for the distributed case, you only have to give the size as an argument.

LeoGaspard replied:
This is implemented and move to the appropriate file.

awietek commented on xdiag/basis/electron/apply/apply_number_number.hpp:
Looks good

awietek commented on xdiag/basis/electron_distributed/apply/apply_number_number.hpp:
looks good

awietek commented on xdiag/basis/electron_distributed/apply/apply_terms.hpp:
```diff
@@ -11,6 +11,7 @@
 #include <xdiag/basis/electron_distributed/apply/apply_raise_lower.hpp>
 #include <xdiag/basis/electron_distributed/apply/apply_szsz.hpp>
 #include <xdiag/basis/electron_distributed/apply/apply_u.hpp>
+#include <xdiag/basis/apply_identity_distributed.hpp>
```
please follow the suggestions iin apply_identity_distributed.hpp. Move to apply_identity.hpp

awietek commented on xdiag/basis/electron_distributed/apply/apply_terms.hpp:
```diff
@@ -59,6 +60,18 @@ void apply_terms(OpSum const &ops, basis_t const &basis_in,
+        electron_distributed::apply_ndn_ndn<coeff_t>(
+            cpl, op, basis_in, vec_in.memptr(), vec_out.memptr());
+    } else if (type == "Id") {
+        apply_identity_distributed<coeff_t>( 
```
Change call as suggested

awietek commented on xdiag/operators/logic/compilation.cpp:
```diff
@@ -194,6 +194,12 @@ OpSum compile_electron(OpSum const &ops) try {
+    } else if ((type == "NupNup") && (op[0] == op[1])) {
+      ops_final += cpl * Op("Nup", op[0]);
+    } else if ((type == "NdnNdn") && (op[0] == op[1])) {
+      ops_final += cpl * Op("Ndn", op[0]);
```
Here you should add the operator "NdnNup". something like:

```c++
else if (type == "NdnNup") {
      ops_final += cpl * Op("NupNdn", {op[0], op[1]});
    }
```

LeoGaspard replied:
This was implemented but at the previous step involving `ops_compiled` so that in the case of `op[0] == op[1]`, the last simplification step for `NupNdn` can be applied. 
I also added tests for this operator.

awietek commented on xdiag/operators/logic/types.hpp:
```diff
@@ -15,13 +15,14 @@ inline const std::vector<std::string> known_types = {
     "Cup",   "Cdagdn",          "Cdn",      "HubbardU",   "Ntot",   "Nup",
     "Ndn",   "Nupdn",           "NtotNtot", "NupdnNupdn", "tJSzSz", "tJSdotS",
-    "Matrix"};
+    "Matrix", "NupNdn", "NupNup", "NdnNdn"};
```
Add NdnNup here

awietek commented on xdiag/operators/logic/types.hpp:
```diff
@@ -15,13 +15,14 @@ inline const std::vector<std::string> known_types = {
     "Cdagdn", "Cdn",      "HubbardU",   "Ntot",   "Nup",    "Ndn",
-    "Nupdn",  "NtotNtot", "NupdnNupdn", "tJSzSz", "tJSdotS"};
+    "Nupdn",  "NtotNtot", "NupdnNupdn", "tJSzSz", "tJSdotS",
+    "NupNdn", "NupNup", "NdnNdn"};
```
Add NdnNup here

awietek commented on xdiag/operators/logic/types.hpp:
```diff
@@ -49,7 +50,11 @@ inline const std::map<std::string, int64_t> _nsites_of_type = {
+    {"NupNdn", 2},
+    {"NupNup", 2},
+    {"NdnNdn", 2},
+};
```
Add NdnNup here

awietek commented on xdiag/operators/logic/valid.cpp:
```diff
@@ -25,7 +25,8 @@ void check_valid(Op const &op) try {
                (type == "NtotNtot") || (type == "NupdnNupdn") ||
-               (type == "tJSzSz") || (type == "tJSdotS")) {
+               (type == "tJSzSz") || (type == "tJSdotS") || (type == "NupNdn") ||
+               (type == "NupNup") || (type == "NdnNdn")) {
```
Add NdnNup here

awietek commented on xdiag/operators/logic/order.cpp:
We should take care of the case that an NdnNup operator gets properly ordered to NupNdn to compare it.

LeoGaspard replied:
Here I simply told the program not to order `NdnNup`, as it is transformed into `NupNdn` during the compilation step. Do you think something better should be done ?

awietek commented on docs/documentation/operators/operator_types.md:
```diff
@@ -19,6 +19,7 @@ Generic operators in XDiag are represented as [OpSum](opsum.md) objects made up
 | `Nup`             | A number operator for an $\uparrow$ spin $n_{i\uparrow}$                                                                                                                   | 1            | tJ, Electron, tJDistributed, ElectronDistributed                                |
 | `Ndn`             | A number operator for an $\downarrow$ spin $n_{i\downarrow}$                                                                                                               | 1            | tJ, Electron, tJDistributed, ElectronDistributed                                |
 | `Ntot`            | A number operator $n_i = n_{i\uparrow} + n_{i\downarrow}$                                                                                                                  | 1            | tJ, Electron, tJDistributed, ElectronDistributed                                |
+| `Nhtot`           | A number operator for holes $n^H_i = 2 - n_{i\uparrow} - n_{i\downarrow}$                                                                                                  | 1            | tJ, Electron, tJDistributed, ElectronDistributed                                |
```
Remove Nhtot here

awietek reviewed (changes requested):
Thanks, this is almost ready to be merged. I have two suggestions:

* The identity operator implementation should be changed, see annotations
*  The NdnNup operator should be added in addition to the NupNdn operator, by compilation.

awietek commented:
This looks good now and I'll merge. I will add one thing about ordering "NupNdn" though. Thank you for your contribution!
